Port to CollabWrapper

Saumya-Mishra9129 opened the pull request:
- [x] Port to Collabwrapper

- [x] Test collaboration

There are the things I tried while testing collaboration -

- joiner is able to join the sharer.
- Clicking on Start a new game by sharer , starts a new game at joiner. Also only sharer is able to start a new game, joiner cannot.
- Same grid is shared between sharer and joiner ( same colours also).
- If joiner recognises a pattern and clicks on True or False buttons , game state will be updated in sharer and vice-versa.
- Explore some examples is not shared between sharer and joiner.

There are 10 levels. Game will start from 1st level , if you give 5 correct answers, you will be updated to next level. These are the patterns for each level.

- Level 1: Center dot is Red
- Level 2: Monochrome
- Level 3: Max chroma
- Level 4: lighter as it moves up
- Level 5: chroma increases to center
- Level 5: diagonal is True
- Level 6: equiluminent
- Level 7: complementary
- Level 8: hue increase down
- Level 9: interlocking hues
- Level 10: checkerboard

Collaboration test is successful. Please review and merge. @quozl @chimosky .

Saumya-Mishra9129 committed:
Port to CollabWrapper

dbus.ExportedGObject class used by ChatTube has gone away
in Python 3, so in Python 3 we must port away from
ExportedGObject.

Telepathy Tubes is deprecated.  It is not essential to port away from
them, but it should be done eventually.

CollabWrapper does not depend on either.

Add CollabWrapper from https://github.com/sugarlabs/collabwrapper

chimosky commented on ColorDeductoActivity.py:
```diff
@@ -84,7 +73,12 @@ def __init__(self, handle):
         self._sharing = False
         self._initiating = False
-        self._setup_presence_service()
+        self.connect('shared', self._shared_cb)
+        self.connect('joined', self._joined_cb)
+        self._collab = CollabWrapper(self)
+        self._collab.connect('message', self._message_cb)
+        self._collab.connect('joined', self._joined_cb)
```
Can you keep the signals in the same stack? Why connect both from sugar and collabwrapper to the same callbacks?

Saumya-Mishra9129 replied:
Done in 2d5baeb.

quozl replied:
Thanks.  They are the same signal name, but they have different meaning, so they aren't the same callbacks.  One indicates readiness as far as sugar3.activity is concerned, the other indicates as far as CollabWrapper is concerned.

chimosky replied:
Yes both signal names have different meaning and get invoked differently but the same callback gets executed when either is invoked.

chimosky commented on ColorDeductoActivity.py:
```diff
@@ -84,7 +73,12 @@ def __init__(self, handle):
-        self._setup_presence_service()
+        self.connect('shared', self._shared_cb)
+        self.connect('joined', self._joined_cb)
+        self._collab = CollabWrapper(self)
```
Since this was also used in another file, you can make it an instance variable rather it be a private one.

Saumya-Mishra9129 replied:
Yeah I will try this.

chimosky commented:
Tested, collaboration worked as expected.

Saumya-Mishra9129 committed:
Remove the restriction while collaborating

Only sharer can share a new game

Saumya-Mishra9129 commented:
> Tested, collaboration worked as expected.

Thanks

chimosky commented:
191538a is supposed to "Remove the restriction while collaborating" and that's supposed to mean the joiner can also share a new game but in `_joined_cb`, `False` is passed to `after_share_join` which sets `self._initiating` to that value and in `_new_game_cb`, only the initiator can share the game state, that kinda defeats the purpose of the commit.

chimosky commented on ColorDeductoActivity.py:
```diff
@@ -73,8 +73,19 @@ def __init__(self, handle):
-        self.connect('shared', self._shared_cb)
-        self.connect('joined', self._joined_cb)
+
+        def on_activity_joined_cb(me):
+            logging.debug('activity joined')
+
+        self.connect('joined', on_activity_joined_cb)
+
+        def on_activity_shared_cb(me):
+            logging.debug('activity shared')
+            self.status.set_label(_('Wait for other players to join.'))
+            self.after_share_join(True)
+
+        self.connect('shared', on_activity_shared_cb)
```
This is the same as before but only this time they're different methods, the idea was if you're using collabwrapper for collaboration, keep those signal connections to collabwrapper and not use both sugar and collabwrapper for the calls.

Saumya-Mishra9129 replied:
@chimosky I am bit confused. Can we not just use collabwrapper joined signal ? As CollabWrapper is used for collaboration, we should just call the joined event of collabwrapper and not the one from the sugar3.activity.activity.Activity class.

chimosky replied:
Yes as you're using collabwrapper, the signal calls should be from collabwrapper and not from `sugar3.activity.activity.Activity`.

Saumya-Mishra9129 committed:
Remove super and use activity.Activity's __init__

we support multiple inheritance in activity, and super is for single inheritance only. We cannot use super in the subclasses if superclass doesn't use it

Saumya-Mishra9129 committed:
Remove call to activity joined signal

As CollabWrapper is used here for collaboration, use the joined event of collabwrapper only and not one from the sugar3.activity.Activity class.

Saumya-Mishra9129 committed:
Fix : Joiner can also initiate a new game

Saumya-Mishra9129 commented:
I tried removing the activity's joined signal. Works fine. @quozl @chimosky Please review.

chimosky commented on ColorDeductoActivity.py:
```diff
@@ -343,7 +340,7 @@ def _message_cb(self, collab, buddy, msg):
 
     def _send_new_game(self):
-        ''' Send a new game message to all players (only sharer sends grids) '''
+        ''' Send a new game message to all players '''
```
Your commit message says something, why remove the doc string that reiterates that?

chimosky commented on ColorDeductoActivity.py:
```diff
@@ -356,7 +353,7 @@ def _receive_new_game(self, payload):
         self.status.set_label(_('Playing level %d') % (self.level + 1))
 
     def _send_new_grid(self):
-        ''' Send a new grid to all players (only sharer sends grids) '''
```
Same issue here.

chimosky commented:
As you intend to make both sharer and joiner be able to share new games, I don't see a need for `self.initiating` anymore, do you?
